wp-content/themes/core/components/navigation/Navigation_Controller.php

```diff
@@ -0,0 +1,82 @@
+<?php declare(strict_types=1);
+
+namespace Tribe\Project\Templates\Components\navigation;
+
+use Tribe\Libs\Utils\Markup_Utils;
+use Tribe\Project\Nav_Menus\Menu;
+use Tribe\Project\Templates\Components\Abstract_Controller;
+
+class Navigation_Controller extends Abstract_Controller {
+
+	public const LOCATION         = 'location';
+	public const CLASSES          = 'classes';
+	public const ATTRS            = 'attrs';
+	public const NAV_LIST_CLASSES = 'nav_list_classes';
+	public const NAV_MENU_ARGS    = 'nav_menu_args';
+
+	private const DEFAULT_NAV_MENU_ARGS = [
+		'container'   => false,
+		'menu_class'  => '',
+		'menu_id'     => '',
+		'depth'       => 1,
+		'items_wrap'  => '%3$s',
+		'fallback_cb' => false,
+	];
+
+	private string $location;
+	private array $classes;
```

Comment:
If you wanted to save me some time when moving to the PHPCS v2 rules you can add the docblock for the type that each array is going to contain, so some IDE's will warn the user when using the using incorrect data, e.g.

```php
	/**
	 * @var string[]
	 */
```

Reply:
Sounds good. Related: for a key => value array (such as attributes), is there a proper style for hinting the type required for both the key and the value? (As supposed to the `string[]` where just the value type is hinted.)

Reply:
You bet! Here are some examples

-   `int[]`: an array with only INT values -- [1, 4, 6, 8, 9, ...]
-   `array<int, int>`: an array with only INT values -- [4 => 1, 8 => 4, 12 => 6, ...]
-   `string[]`: an array with only STRING values -- [„foo", „bar", ...]
-   `array<int, string>`: an array with only STRING values -- [4 => „foo", 8 => „bar", ...]
-   `Order[]`: an array with only „Order"-Object values -- [Order, Order, ...]
-   `array<int|string, Order>`: an array with INT or STRING as key and „Order"-Object values -- [4 => Order, 'foo' => Order, ...]
-   `array<int|string, mixed>`: an array with INT or STRING as key and mixed as values -- [1 => 1, 4 => „foo", 6 => \stdClass, ...]
-   `array<int, array<int, string>>`: an array with INT as key and and an array (with INT as key and string as value) as values -- [1 => [1 => „foo"], 4 => [1 => 4], ...]
-   `array<int, string[]>`: an array with INT as key and and an array (with INT as key and string as value) as values -- [1 => [„foo", „lall"], 4 => [„öäü", „bar"], ...]
-   `array{output: string, debug: string}`: an array with the key „output" and „debug", the values are STRING values -- [‚output' => ‚foo', ‚debug' => ‚bar']
-   `array<int, array{output: string, debug: string}>`: an array with the key „output" and „debug", the values are STRING values -- [1 => [‚output' => ‚foo', ‚debug' => ‚bar'], 3 => [‚output' => ‚foo', ‚debug' => ‚bar'], ...]

[sauce](https://dev.to/suckup_de/modern-phpdoc-annotations-11d4)

Or some other basic associative arrays:

- `string[][]`
- `int[][]`

etc...